Add option to trigger outgoing webhook if first word starts with trigger word

Rudloff opened the pull request:
#### Summary
This patch adds a new option when adding an outgoing webhook:
![mattermost webhook](https://cloud.githubusercontent.com/assets/840125/16903863/3bc2e558-4c8a-11e6-918f-ddbda5ee896d.png)
The first option is the current behavior.
The second one triggers the hook if the first word of a post starts with the trigger word (e.g if your trigger word is _foo_, _foobar_ will trigger it).

#### Ticket Link
https://github.com/mattermost/platform/issues/3574

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] Has driver changes that have been merged and package.json updated
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [ ] Includes text changes and localization files updated
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

jwilander commented:
@jasonblais can you take a look?

mattermod commented:
Spinmint test server being created...

mattermod commented:
Spinmint test server created.

Link: http://i-d15ead45.spinmint.com:8065/pr3611

Email: `test@test.com`

Email2: `test2@test.com`

Password: `passwd`

Instance ID: i-d15ead45

jasonblais commented:
Thanks @Rudloff for your submission! Love your initiative here

Sorry for the delay in the review process, currently testing different cases and will get back to you on this tomorrow.

Thanks again!!

jasonblais commented:
Thanks @Rudloff once more -- I was having technical difficulties getting one of our webhooks set up, but it's hooked now and I was able to test it successfully. Thank you for your patience -- I'll be able to test future updates on this much sooner.

I tested your pull request on Windows 10, Chrome, and set up outgoing webhooks using a [Giphy Integration Service for Mattermost](https://github.com/numberly/mattermost-integration-giphy). 

1) If I select `Trigger when: first word matches trigger word exactly`, everything works as expected

2) If I select `Trigger when: first word starts with trigger word`, with trigger word `gif`
 - sending a message `gif hello` sends a response and I receive a giphy as expected
 - sending a message `gifhello hello`, `giff hello` or other combinations provides no results

Hence, it appears I'm unable to get this to work as expected? I've attached a screenshot of two webhooks I've tried. You may also use the test server above, if you'd like to try it out.

![image](https://cloud.githubusercontent.com/assets/13119842/17072022/47f81c6a-5033-11e6-8daa-0091384a91c6.png)

3) Will this change apply if I have specified multiple trigger words? For instance, if I had `gif` and `giphy`, with the option `Trigger when first word starts with trigger word`, will posts starting with EITHER `gif` OR `giphy` trigger the webhook?

4) For the UI text, propose making slight tweaks for it:
 - `Trigger when` --> `Trigger When`
 - `First word matches trigger word exactly` --> `First word matches a trigger word exactly`
 - `First word starts with trigger word` --> `First word starts with a trigger word`

![image](https://cloud.githubusercontent.com/assets/13119842/17072146/251b84d8-5034-11e6-9c4a-f1dfbe0950bb.png)

Rudloff commented:
Sorry, it looks like I made a mistake with my last push and some of the code was missing :/
I updated my PR and it should work correctly now.
(I also updated the strings as you suggested.)

To answer 3: yes it should work with either word if you add multiple trigger words.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Spinmint test server being created...

mattermod commented:
Spinmint test server created.

Link: http://i-5a31f5ce.spinmint.com:8065/pr3611

Email: `test@test.com`

Email2: `test2@test.com`

Password: `passwd`

Instance ID: i-5a31f5ce

jasonblais commented:
+1 for PM approval

Tested the following cases successfully on Windows 10, Chrome (trigger words contain gif)

 - messages triggering webhook (each one followed with the word 'hey'): gif, gif#, gif-hey, gifhey, gif*, gif:, gif//
 - messages not triggering the webhook: #gif, agif, .gif, # gif, hey gif

Tested above cases in the following situations:
 - when only one trigger word specified, as well as when multiple trigger words are specified.
 - when `First word matches a trigger word exactly` and when `First word starts with a trigger word` options chosen

No issues found, big thank you @Rudloff! This is a nice improvement for outgoing webhooks :+1:

enahum commented:
Hi @Rudloff thanks for the PR, can you please add the relevant text to the correspondent `en.json` files?

Rudloff commented:
@enahum I updated my PR. Is it OK like this?

enahum commented on webapp/components/integrations/components/add_outgoing_webhook.jsx:
```diff
@@ -300,6 +309,41 @@ export default class AddOutgoingWebhook extends React.Component {
+                                htmlFor='triggerWords'
+                            >
+                                <FormattedMessage
+                                    id='add_outgoing_webhook.triggerWordsTriggerWhen'
```
Thanks @Rudloff for your quick response.

This one is still missing in `en.json` my friend.

Rudloff replied:
Oh sorry, I forgot it. I updated the PR.

enahum commented:
Thanks +1

crspeller commented on api/post.go:
```diff
@@ -384,11 +384,19 @@ func handleWebhookEvents(c *Context, post *model.Post, team *model.Team, channel
-			if len(hook.TriggerWords) == 0 || hook.HasTriggerWord(firstWord) {
+			if len(hook.TriggerWords) == 0 {
+				relevantHooks = append(relevantHooks, hook)
+			} else if hook.TriggerWhen == 0 && hook.HasTriggerWord(firstWord) {
```
Can we get some consents for these values? Maybe have a function that encapsulates this logic? This whole section is really hard to understand.

Rudloff replied:
I think we can refactor like that:
```go
for _, hook := range hooks {
	if hook.ChannelId == post.ChannelId || len(hook.ChannelId) == 0 {
		if hook.ChannelId == post.ChannelId && len(hook.TriggerWords) == 0 {
			relevantHooks = append(relevantHooks, hook)
		} else if hook.TriggerWhen == 0 && hook.HasTriggerWord(firstWord) {
			relevantHooks = append(relevantHooks, hook)
		} else if hook.TriggerWhen == 1 && hook.TriggerWordStartsWith(firstWord) {
			relevantHooks = append(relevantHooks, hook)
		}
	}
}
```
Would that be OK for you?

crspeller replied:
Better, but let's get some constants for the TriggerWhen `0` and `1` values.

crspeller commented on model/outgoing_webhook.go:
```diff
@@ -204,3 +210,17 @@ func (o *OutgoingWebhook) HasTriggerWord(word string) bool {
 	return false
 }
+
+func (o *OutgoingWebhook) TriggerWordStartsWith(word string) bool {
```
Unit test?

Rudloff replied:
I added a basic unit test. Is that enough?

Rudloff commented on api/post.go:
```diff
@@ -383,12 +388,14 @@ func handleWebhookEvents(c *Context, post *model.Post, team *model.Team, channel
 
 	relevantHooks := []*model.OutgoingWebhook{}
 	for _, hook := range hooks {
```
@crspeller Is it OK like this?

crspeller replied:
@Rudloff Yes, after you fix the gofmt error.

Rudloff replied:
Whoops, sorry I forgot to run `make check-server-style`.
It should be good now.

crspeller commented:
+1
